akka-http-core/src/main/scala/akka/http/engine/client/ConnectionPool.scala

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.engine.client
+
+import java.net.InetSocketAddress
+import scala.concurrent.{ Promise, Future }
+import akka.event.LoggingAdapter
+import akka.actor._
+import akka.stream.{ OperationAttributes, FlowMaterializer }
+import akka.stream.scaladsl._
+import akka.http.model._
+import akka.http.util._
+import akka.http.Http
+
+private[http] object ConnectionPool {
+  import HostConnectionPoolGateway._
+
+  case class PoolRequest(request: HttpRequest, responsePromise: Promise[HttpResponse])
+  case object BeginShutdown extends DeadLetterSuppression
+
+  /**
+   * Starts a new host connection pool and returns its gateway actor.
+   *
+   * A host connection pool for a given [[HostConnectionPoolSetup]] is a running stream, whose outside interface is
+   * provided by its [[PoolGateway]] actor.
+   * This gateway actor accepts [[PoolRequest]] messages and completes their `responsePromise` whenever the respective
+   * response has been received (or an error occurred).
+   * A gateway actor can be orderly shut down by sending it a [[BeginShutdown]] message.
+   *
+   * A running host connection pool automatically shuts down itself after the configured idle timeout.
+   */
+  def startHostConnectionPool(hcps: HostConnectionPoolSetup)(implicit system: ActorSystem,
+                                                             fm: FlowMaterializer): Gateway = {
+    import hcps._
+    import setup._
+    val connectionFlow = Http().outgoingConnection(host, port, None, options, settings.connectionSettings, log)
+    val poolFlow = newPoolFlow(connectionFlow, new InetSocketAddress(host, port), settings, log)
+    val gateway = new Gateway(new PoolGateway(poolFlow, hcps, _))
+    gateway.actorRef() // trigger eager start of pool gateway (and pool)
+    gateway
+  }
+
+  /*
+    Pool Flow Stream Setup
+    ======================
+                                               +-------------------+                             
+                                               |                   |                             
+                                        +----> | Connection Slot 1 +---->                        
+                                        |      |                   |    |                        
+                                        |      +---+---------------+    |                        
+                                        |          |                    |                        
+                       +-----------+    |      +-------------------+    |      +---------------+
+      RequestContext   |           +----+      |                   |    +----> |               |  ResponseContext
+    +----------------> | Conductor |---------> | Connection Slot 2 +---------> | responseMerge +------------------>
+                       |           +----+      |                   |    +----> |               |
```

Comment:
(or the github formatter screws this up—the view definitely does, but I copied it into a terminal window)